File: SurvataObjcDemo/ViewController.m
//
//  ViewController.m
//  SurvataObjcDemo
//
//  Created by Rex Sheng on 2/22/16.
//  Copyright © 2016 InteractiveLabs. All rights reserved.
//

#import "ViewController.h"
@import Survata;
@import CoreLocation;

@interface ViewController () <CLLocationManagerDelegate>
@property (nonatomic, weak) IBOutlet UIView *surveyMask;
@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *surveyIndicator;
@property (nonatomic, weak) IBOutlet UIButton *surveyButton;
@property (nonatomic) SVSurvey *survey;
@end

@implementation ViewController
{
	CLLocationManager *locationManager;
}

- (void)viewWillAppear:(BOOL)animated {
	[super viewWillAppear:animated];
	if ([CLLocationManager authorizationStatus] == kCLAuthorizationStatusNotDetermined) {
		locationManager = [CLLocationManager new];
		locationManager.delegate = self;
		[locationManager requestWhenInUseAuthorization];
	} else {
		[self createSurvey];
	}
}

- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
	if (status == kCLAuthorizationStatusAuthorizedWhenInUse) {
		[self createSurvey];
	}
	if (status != kCLAuthorizationStatusNotDetermined) {
		locationManager = nil;
	}
}

- (void)showFull {
	self.surveyMask.hidden = YES;
}

- (void)showSurveyButton {
	self.surveyMask.hidden = NO;
	self.surveyButton.hidden = NO;
	[self.surveyIndicator stopAnimating];
}

- (IBAction)startSurvey {
	__weak __typeof(self) weakSelf = self;
	[survey createSurveyWall:self completion:^(enum SVSurveyResult result) {
		if (result == SVSurveyResultCompleted) {
			[weakSelf showFull];
		} else {
			[weakSelf showSurveyButton];
		}
	}];
}

- (void)createSurvey {
	if (survey) return;
	SVSurveyOption *option = [[SVSurveyOption alloc] initWithPublisher:@"survata-test"];
	option.preview = @"46b140a358cd4fe7b425aa361b41bed9";
	__weak __typeof(self) weakSelf = self;
	survey = [[SVSurvey alloc] initWithOption:option];
	[survey create:^(enum SVSurveyAvailability availability) {
		__strong __typeof(weakSelf) strongSelf = weakSelf;
		if (!strongSelf) return;
		if (availability == SVSurveyAvailabilityAvailable) {
			[weakSelf showSurveyButton];
		} else {
			[weakSelf showFull];
		}
	}];
}

@end

File: SurvataDemo/ViewController.swift
//
//  ViewController.swift
//  Demo
//
//  Created by Rex Sheng on 2/11/16.
//  Copyright © 2016 Survata. All rights reserved.
//

import Survata
import CoreLocation

class ViewController: UIViewController, CLLocationManagerDelegate {
	
	@IBOutlet weak var surveyMask: UIView!
	@IBOutlet weak var surveyIndicator: UIActivityIndicatorView!
	@IBOutlet weak var surveyButton: UIButton!
	
	var created = false
	var locationManager: CLLocationManager!
	override func viewWillAppear(animated: Bool) {
		super.viewWillAppear(animated)
		if CLLocationManager.authorizationStatus() == .NotDetermined {
			locationManager = CLLocationManager()
			locationManager.delegate = self
			locationManager.requestWhenInUseAuthorization()
		} else {
			createSurvey()
		}
	}
	
	func locationManager(manager: CLLocationManager, didChangeAuthorizationStatus status: CLAuthorizationStatus) {
		if status == .AuthorizedWhenInUse {
			createSurvey()
		}
		if status != .NotDetermined {
			locationManager = nil
		}
	}
	
	override func canBecomeFirstResponder() -> Bool {
		return true
	}
	
	override func motionEnded(motion: UIEventSubtype, withEvent event: UIEvent?) {
		if motion == .MotionShake {
			let controller = UIAlertController(title: "Reset Demo?", message: nil, preferredStyle: .Alert)
			let option = UIAlertAction(title: "Reset", style: .Destructive) {[weak self] _ in
				guard let window = self?.view.window, storyboard = self?.storyboard else { return }
				NSHTTPCookieStorage.sharedHTTPCookieStorage().removeCookiesSinceDate(NSDate(timeIntervalSince1970: 0))
				window.rootViewController = storyboard.instantiateInitialViewController()
			}
			controller.addAction(option)
			controller.addAction(UIAlertAction(title: "Cancel", style: .Cancel, handler: nil))
			self.presentViewController(controller, animated: true, completion: nil)
		}
	}
	
	func showFull() {
		surveyMask.hidden = true
	}
	
	func showSurveyButton() {
		surveyMask.hidden = false
		surveyButton.hidden = false
		surveyIndicator.stopAnimating()
	}
	
	@IBAction func startSurvey() {
		Survey.presentFromController(self, option: SurveyOption()) {[weak self] result in
			switch result {
			case .Completed:
				self?.showFull()
			default:
				self?.showSurveyButton()
			}
		}
	}
	
	func createSurvey() {
		if created { return }
		print("Survey.create...")
		Survey.create("survata-test") {[weak self] result in
			self?.created = true
			switch result {
			case .Available:
				self?.showSurveyButton()
			default:
				self?.showFull()
			}
		}
	}
}